### ISSUE TITLE
Load History not working always?

### ISSUE THREAD

**Issue by psmaker** [RunestoneInteractive/RunestoneServer#1170]
Course: LHSIntroCS_2018

This isn't a bug, but rather some behavior I've noticed two days not in my class where student's code that they have written is not showing when they press the ``Load History`` button. But, then by the end of the class it was working.

Today, same issue - the ``Load History`` button was not working for multiple students. Then about 10 minutes later, it is working.  The other odd thing is that it is with the same class - not with my other 2 classes. So, I'm wondering if it has something to do with the time of day when they are encountering it - perhaps other people using the server at this same time? (Around 9:30am PST).  My afternoon classes have not seen this problem.

Or, I'm wondering if it has something to do with the size of my classroom and the size of the database. I have about 110 students.

**Comment by bnmnetp** [comment]
That is definitely during the busiest few hours of the day.

To be clear the Save and Run appears to be working, but the load history is failing.  Is it really failing or is it just really sluggish?

If you could check the javascript console the next time you experience this and pass along any error messages you see that would be super helpful.


Runestone continues to grow in popularity, and I may need to look at scaling up the hardware again...

Brad

**Comment by bnmnetp** [comment]
I looked into this a bit and realized that the average time to service the API call to load the history had grown to over 3 seconds!  On top of that the gethist API call is one of the more common calls in use across the books.

Upon further investigation I found that the database table that contains the history had no index!  Adding the appropriate index has reduced this time to a few hundred milliseconds.  This should help overall site performance as well as hopefully eliminating the problems you are seeing with your class.
